core/src/databases/database.rs

```diff
@@ -132,85 +130,101 @@ impl Table {
-    ) -> Result<Option<TableSchema>> {
-        match &self.schema_stale_at {
-            Some(_) => {
-                let schema = self.compute_schema(databases_store).await?;
-                store
-                    .update_table_schema(
+    ) -> Result<()> {
+        if self.table_type()? == TableType::LOCAL {
+            // Invalidate the databases that use the table.
+            try_join_all(
+                (store
+                    .find_databases_using_table(
                         &self.project,
                         &self.data_source_id,
                         &self.table_id,
-                        &schema,
+                        HEARTBEAT_INTERVAL_MS,
                     )
-                    .await?;
-                self.schema = Some(schema.clone());
-                Ok(Some(schema.clone()))
-            }
-            None => Ok(self.schema.clone()),
+                    .await?)
+                    .into_iter()
+                    .map(|db| {
+                        let store = store.clone();
+                        async move {
+                            db.invalidate(store).await?;
+                            Ok::<_, anyhow::Error>(())
+                        }
+                    }),
+            )
+            .await?;
+
+            // Delete the table rows.
+            databases_store.delete_table_rows(&self.unique_id()).await?;
         }
+
+        store
+            .delete_table(&self.project, &self.data_source_id, &self.table_id)
+            .await?;
+
+        Ok(())
     }
 
-    pub async fn delete(
+    pub async fn update_parents(
         &self,
         store: Box<dyn Store + Sync + Send>,
-        databases_store: Box<dyn DatabasesStore + Sync + Send>,
+        parents: Vec<String>,
     ) -> Result<()> {
-        // Invalidate the databases that use the table.
-        try_join_all(
-            (store
-                .find_databases_using_table(
-                    &self.project,
-                    &self.data_source_id,
-                    &self.table_id,
-                    HEARTBEAT_INTERVAL_MS,
-                )
-                .await?)
-                .into_iter()
-                .map(|db| {
-                    let store = store.clone();
-                    async move {
-                        db.invalidate(store).await?;
-                        Ok::<_, anyhow::Error>(())
-                    }
-                }),
-        )
-        .await?;
-
-        // Delete the table rows.
-        databases_store.delete_table_rows(&self.unique_id()).await?;
-
         store
-            .delete_table(&self.project, &self.data_source_id, &self.table_id)
+            .update_table_parents(
+                &self.project,
+                &self.data_source_id,
+                &&self.table_id,
+                &parents,
+            )
             .await?;
-
         Ok(())
     }
+}
+
+pub struct LocalTable<'a> {
+    pub table: &'a mut Table,
```

Comment:
I think you can just copy here. It will make our lives easier.

Reply:
Well it works pretty well like this without any copying no ? Makes everything cleaner and prevents potential bugs IMO

Reply:
I like the idea of wrapping an existing table in a struct that "extends" its capabilities. It makes sense to me that once you do this wrapping, the `LocalTable` owns the underlying table object and that they have the same lifetime

Reply:
This means that you can't pull your local table easily in another context without making sure that Table survives the move. It works now but can make things more complex in future situations. The cost of owning a Table  here is potentially copying a smallish object but in most cases it will likely be a move so there would be no copying at all. And I don't think we will have cases where we want to use the Table independently of the wrapping LocalTable?